libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -790,10 +807,65 @@ bool ContractCompiler::visit(InlineAssembly const& _inlineAssembly)
 			_assembly.appendInstruction(Instruction::POP);
 		}
 	};
-	solAssert(_inlineAssembly.annotation().analysisInfo, "");
+
+	yul::Block const* code = &_inlineAssembly.operations();
+	yul::AsmAnalysisInfo* analysisInfo = &*_inlineAssembly.annotation().analysisInfo;
+
+	yul::Object object;
+
+	// The optimiser cannot handle external references
+	if (
+		m_optimiserSettings.runYulOptimiser &&
+		_inlineAssembly.annotation().externalReferences.empty()
+	)
+	{
+		// Create a modifiable copy of the code and analysis
+		object.code = make_shared<yul::Block>(
+			yul::ASTCopier().translate(_inlineAssembly.operations())
+		);
+		code = &*object.code;
```

Comment:
I think it would be better to assign code and analysisInfo only at the very end of the if block.